File: ch4/akkademy-db-client-java/src/main/java/com/akkademy/clientactor/FSMClientActor.java
package com.akkademy.clientactor;

import akka.actor.AbstractFSM;
import akka.actor.ActorSelection;
import com.akkademy.messages.Connected;
import com.akkademy.messages.Request;

import java.util.LinkedList;

import static com.akkademy.clientactor.State.*;

enum State{
    DISCONNECTED,
    CONNECTED,
    CONNECTED_AND_PENDING
}

class EventQueue extends LinkedList<Request> {}
class FlushMsg{}

/**
 * To revert to disconnected, we could send occassional heartbeat pings
 * and revert by restarting the actor (throwing an exception)
 */

public class FSMClientActor extends AbstractFSM<State, EventQueue>{

    private ActorSelection remoteDb;

    public FSMClientActor(String dbPath){
        remoteDb = context().actorSelection(dbPath);
    }

    {
        startWith(DISCONNECTED, new EventQueue());

        when(DISCONNECTED,
                matchEvent(FlushMsg.class, (msg, container) -> stay())
                        .event(Request.class, (msg, container) -> {
                            remoteDb.tell(new Connected(), self());
                            container.add(msg);
                            return stay();
                        }).event(Connected.class, (msg, container) -> {
                    if(container.size() == 0){
                        return goTo(CONNECTED);
                    }else{
                        return goTo(CONNECTED_AND_PENDING);
                    }
                }));

        when(CONNECTED,
                matchEvent(FlushMsg.class, (msg, container) ->
                        stay()
                )
                        .event(Request.class, (msg, container) -> {
                            container.add(msg);
                            return goTo(CONNECTED_AND_PENDING);
                        }));


        when(CONNECTED_AND_PENDING,
                matchEvent(FlushMsg.class, (msg, container) -> {
                    remoteDb.tell(container, self());
                    container = new EventQueue();
                    return goTo(CONNECTED);
                })
                        .event(Request.class, (msg, container) -> {
                            container.add(msg);
                            return goTo(CONNECTED_AND_PENDING);
                        }));

        initialize();
    }
}

File: ch4/akkademy-db-client-scala/src/main/scala/com/akkademy/clientactor/FSMClientActor.scala
package com.akkademy.clientactor

import akka.actor.FSM
import com.akkademy.clientactor.StateContainerTypes.RequestQueue
import com.akkademy.messages
import com.akkademy.messages._

sealed trait State
case object Disconnected extends State
case object Connected extends State
case object ConnectedAndPending extends State

case object Flush
case object ConnectedMsg

object StateContainerTypes {
  type RequestQueue = List[Request]
}

class FSMClientActor(address: String) extends FSM[State, RequestQueue]{
  private val remoteDb = context.system.actorSelection(address)

  startWith(Disconnected, List.empty[Request])

  when(Disconnected){
    case Event(_: messages.Connected, container: RequestQueue) => //If we get back a ping from db, change state
      if (container.headOption.isEmpty)
        goto(Connected)
      else
        goto(ConnectedAndPending)
    case Event(x: Request, container: RequestQueue) =>
      remoteDb ! new messages.Connected //Ping remote db to see if we're connected if not yet marked online.
      stay using (container :+ x) //Stash the msg
    case x =>
      println("uhh didn't quite get that: " + x)
      stay()
  }

  when (Connected) {
    case Event(x: Request, container: RequestQueue) =>
      goto(ConnectedAndPending) using(container :+ x)
  }

  when (ConnectedAndPending) {
    case Event(Flush, container) =>
      remoteDb ! container;
      goto(Connected) using Nil
    case Event(x: Request, container: RequestQueue) =>
      stay using(container :+ x)
  }

  initialize()
}
